Add partial OpenSharedHandle support

MichelVachon opened the pull request:
This commit is about issue #1199.

Thanks

MichelVachon committed:
Add partial OpenSharedHandle support

Add partial (as in write only, no initial data) D3D12 OpenSharedHandle support

MichelVachon commented:
I will update the source to fix clang and gcc, hold on

MichelVachon commented:
I will update the source to fix clang and gcc, hold on

MichelVachon committed:
Fix clang and gcc

MichelVachon committed:
yet another formatting fix

MichelVachon committed:
haha damn

baldurk commented on renderdoc/driver/d3d12/d3d12_device_wrap.cpp:
```diff
@@ -2134,8 +2134,89 @@ void WrappedID3D12Device::CopyDescriptorsSimple(UINT NumDescriptors,
+        SCOPED_SERIALISE_CHUNK(D3D12Chunk::Device_CreateCommittedResource);
+        Serialise_CreateCommittedResource(ser, &heapProperties, heapFlags, &desc,
+                                          InitialResourceState, pOptimizedClearValue, riid,
+                                          (void **)&wrapped);
```
I see what you mean about re-using `CreateCommittedResource` now :smile:. I think for user clarity and to prevent being too tied into the implementation of `Serialise_CreateCommittedResource` I'd rather have a separate chunk type & `Serialise_OpenSharedHandle` for this call. Otherwise it becomes a bit strange to deal with for different types of shared resources like fences/heaps.

Similar to how the D3D11 one works, it serialises the type parameter and resource ID (in this case the `REFIID riid` type would be better than the D3D11's `IdentifyTypeByPtr` type), then checks on the type and serialises any further parameters that it queries out of the interface, the same way as you've queried them out here.

baldurk commented on renderdoc/driver/d3d12/d3d12_device_wrap.cpp:
```diff
@@ -2134,8 +2134,89 @@ void WrappedID3D12Device::CopyDescriptorsSimple(UINT NumDescriptors,
+      const D3D12_RESOURCE_DESC desc = real->GetDesc();
+
+      D3D12_RESOURCE_STATES InitialResourceState = D3D12_RESOURCE_STATE_GENERIC_READ;
+      const D3D12_CLEAR_VALUE *pOptimizedClearValue = nullptr;
```
Style nitpick: prefer `NULL` to `nullptr` please.

baldurk commented on renderdoc/driver/d3d12/d3d12_device_wrap.cpp:
```diff
@@ -2134,8 +2134,89 @@ void WrappedID3D12Device::CopyDescriptorsSimple(UINT NumDescriptors,
+      real->GetHeapProperties(&heapProperties, &heapFlags);
+      const D3D12_RESOURCE_DESC desc = real->GetDesc();
+
+      D3D12_RESOURCE_STATES InitialResourceState = D3D12_RESOURCE_STATE_GENERIC_READ;
```
Is there any documentation saying that shared resources are created in this state, or is it a best-guess? If there's no specific documentation I think `D3D12_RESOURCE_STATE_COMMON` is a better/more compatible default.

baldurk reviewed (changes requested):
This looks overall good but I'd prefer a separate chunk & serialise function rather than re-using `Serialise_CreateCommittedResource`.

FYI the gcc/clang failures weren't linux specific, it's just that those are the most convenient places in the CI to check for code formatting. I ask that code formatting commits be removed from the history and squashed into the relevant commit, so it would be a good idea to do that before making any follow-on changes.

If you don't want to keep any separate commits then if you like this time I can squash the whole PR into a single commit when it's ready to merge, it's up to you.

MichelVachon committed:
OpenSharedHandle chunk

Replace direct usage of CreateCommittedResource serialisation with dedicated OpenSharedHandle serialisation function and chunk

baldurk commented on renderdoc/driver/d3d12/d3d12_common.h:
```diff
@@ -726,6 +726,7 @@ enum class D3D12Chunk : uint32_t
   Device_CreateSampler,
   Device_CopyDescriptors,
   Device_CopyDescriptorsSimple,
+  Device_OpenSharedHandle,
```
Adding the chunk into the middle of the list like this will break opening old captures, as the serialised chunk IDs won't match. Any new chunks should be added to the end of the list.

MichelVachon replied:
I added a compile time assert in d3d12_stringise.cpp that checks the last added chunk type.

baldurk commented on renderdoc/driver/d3d12/d3d12_device_wrap.cpp:
```diff
@@ -2132,10 +2132,183 @@ void WrappedID3D12Device::CopyDescriptorsSimple(UINT NumDescriptors,
+template <typename SerialiserType>
+bool WrappedID3D12Device::Serialise_OpenSharedHandle(SerialiserType &ser, HANDLE NTHandle, REFIID riid, void **ppvObj)
+{
+  ID3D12DeviceChild *devChild = ser.IsWriting() ? (ID3D12DeviceChild *)*ppvObj : NULL;
```
This cast isn't valid because the `void*` coming in is `ID3D12Resource*`. It might work by coincidence but it's a danger for the future. Unfortunately I don't think ther's a better solution than just switching on `riid` up here and casting to the appropriate type to fetch the ID.

baldurk commented on renderdoc/driver/d3d12/d3d12_device_wrap.cpp:
```diff
@@ -2132,10 +2132,183 @@ void WrappedID3D12Device::CopyDescriptorsSimple(UINT NumDescriptors,
+  ID3D12DeviceChild *devChild = ser.IsWriting() ? (ID3D12DeviceChild *)*ppvObj : NULL;
+
+  SERIALISE_ELEMENT_LOCAL(ResourceRIID, riid);
+  SERIALISE_ELEMENT_LOCAL(ResourceId, GetResID(devChild));
```
`ResourceId` is the name of the type so using it as a variable name is rather strange. The convention for serialised arguments is to use something similar to the argument name if the argument isn't serialised in-place, even if the type doesn't match up to what's expected, so for this maybe `pObj` would be good.

baldurk commented:
I think that change is good, thanks. You have some formatting issues in the newest code and I've noted a couple of issues, but otherwise this is about ready.

MichelVachon committed:
Applying requested changes

+ formatting fixes

MichelVachon committed:
Formatting fix

MichelVachon commented:
btw, do what you think is best regarding squashing/merging, since I don't have github locally, I don't seem to have access to that online.

baldurk commented:
No problem. Squashing is likely available in whatever git client you're using under 'rebasing', as that allows you to merge multiple commits together. I'll just squash everything together when merging.

baldurk commented:
Manually squashed and merged.
